storagemarket/deal_execution.go

```diff
@@ -81,10 +77,17 @@ func (p *Provider) execDeal(ctx context.Context, pub event.Emitter, deal *types.
 
 	// Transfer Data
 	if deal.Checkpoint < dealcheckpoints.Transferred {
-		if err := p.transferAndVerify(ctx, pub, deal); err != nil {
+		if err := p.transferAndVerify(dh.transferCtx, pub, deal); err != nil {
+			dh.tdOnce.Do(func() {
+				close(dh.transferDone)
+			})
 			return fmt.Errorf("failed data transfer: %w", err)
 		}
 	}
+	dh.tdOnce.Do(func() {
+		dh.transferDone <- errors.New("transfer already complete")
+		close(dh.transferDone)
+	})
```

Comment:
The code for handling the complete state of a transfer is spread around functions in this file. Would it be possible to add a function to the dealHandler class and call it from here, so as not to expose dealHandler internals?
eg `dh.completed(err)`

Reply:
done.